src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs

```diff
@@ -100,24 +69,15 @@ public CompilationResult Compile(RelativeFileInfo fileInfo, string compilationCo
             }
 
             _logger.GeneratedCodeToAssemblyCompilationStart(fileInfo.RelativePath);
-
             var startTimestamp = _logger.IsEnabled(LogLevel.Debug) ? Stopwatch.GetTimestamp() : 0;
 
             var assemblyName = Path.GetRandomFileName();
-
             var sourceText = SourceText.From(compilationContent, Encoding.UTF8);
-            var syntaxTree = CSharpSyntaxTree.ParseText(
-                sourceText,
-                path: assemblyName,
-                options: _parseOptions);
-
-            var compilation = CSharpCompilation.Create(
-                assemblyName,
-                options: _compilationOptions,
-                syntaxTrees: new[] { syntaxTree },
-                references: CompilationReferences);
-
-            compilation = Rewrite(compilation);
+            var syntaxTree = _compiler.CreateSyntaxTree(sourceText).WithFilePath(assemblyName);
+            var compilation = _compiler
+                .CreateCompilation(assemblyName)
+                .AddSyntaxTrees(syntaxTree);
+            compilation = ExpressionRewriter.Rewrite(compilation);
```

Comment:
We have to invoke it from precompilation. I'll remove the method from here, or just leave it as part of the method but call the new static method instead.